### ISSUE TITLE
[css-sizing] `height: stretch` should just behave as `height: auto`

### ISSUE THREAD

**Issue by tabatkins** [w3c/csswg-drafts#1614]
The motivating example for `height: stretch`'s behavior is "what `height: 100%` should have been" - making the element stretch up to fill its container, even if the container to be filled is further up the ancestor tree (so its direct parent has an indefinite height).  However, this ends up with several weird implications:

* If two descendants both want to be `height: stretch`, what happens?  This'll cause overflow in a confusing manner, I think.
* If no ancestor is definite-height, it ends up hitting the root, which uses the ICB's height. The screen's height is fairly arbitrary here; it doesn't often actually correspond to what you want an element to do, as in [this example](https://lists.w3.org/Archives/Public/www-style/2012Oct/0204.html)

On the other hand, Flex and Grid have useful, reasonable definitions of stretching the height of an element. In the linked example above, for instance, the author could have made the `li` elements into flexboxes as well; then the `a` elements would stretch to full height, exactly as desired. (Rather than stretching to the screen's height.)

So it's probably just not worth trying to do `height: stretch` in some intelligent manner. We should just treat it like `100%` minus margins/border/padding, so it only has a useful effect is the immediate parent is definite-height, and otherwise acts like `auto`.

**Comment by fantasai** [comment]
We could alternately define that `height: stretch` behaves as `align-self: stretch` in the layout modes that have that ability--it's better than defining it as `height: 100%` since (e.g. in Flexbox) it gives better behavior. For the ones that don't we can define it as `height: 100%`. 

(Other thought: since the original definition was written for block layout, we could just keep it for block layout and cut it off at the BFC boundary... if we can figure out what margin collapsing is supposed to do. But I'm less sure that this is useful, since as you say, most of these cases are good candidates for flex/grid.)

**Comment by fantasai** [comment]
Forgot to add: using `height: stretch` allows using `align-self` to specify fallback alignment when you hit the max-height.

**Comment by fantasai** [comment]
As we were drafting text for this resolution, @tabatkins and I noticed that we had discussed what to do in cases where 'align-self: stretch' applies, but not so much where it doesn't.

Our [current draft](https://drafts.csswg.org/css-sizing-4/#stretch-fit-sizing) includes a paragraph defining these cases (where there's no 'align-self: stretch' behavior to hook into, such as the height of block boxes) to match the behavior of '100%', modified to apply to the margin box to effect the desired "fill container" behavior. (This specs the behavior suggested in the earlier https://github.com/w3c/csswg-drafts/issues/1614#issuecomment-316191422 )

In cases where even percentages can't resolve, it falls back to auto.

This new text still avoids the "blow up all the elements up to the nearest definite-height ancestor" behavior that was previously there, which inspired this issue in the first place; instead, it only references the nearest containing block, same as percentage resolution. And while it can still cause overflow in a container with a height:stretch element along with other children, this is no different than 'width: 100%'.
